src/services/dexScaner/UniswapPoolScanner.ts

```diff
@@ -0,0 +1,96 @@
+import {UniswapV3Helper} from '../../blockchains/evm/contracts/UniswapV3Helper.js';
+import {DexPoolScannerService} from './DexPoolScannerService.js';
+
+export type UniswapPoolCreatedEvent = {
+  token0: string;
+  token1: string;
+  fee: bigint;
+  pool: string;
+  anchor: number;
+};
+
+@injectable()
+export class UniswapPoolScanner extends BlockchainScanner {
+  readonly settings: Settings;
+  readonly markerId = 'UniswapPoolScanner';
+
+  @inject(UniswapPoolService) uniswapPoolService!: UniswapPoolService;
+  @inject(DexPoolScannerService) dexPoolScannerService!: DexPoolScannerService;
+  @inject(UniswapV3Factory) uniswapV3Factory!: UniswapV3Factory;
+  @inject(UniswapV3Helper) uniswapV3Helper!: UniswapV3Helper;
+
+  constructor(@inject('Settings') settings: Settings) {
+    super();
+
+    this.settings = settings;
+  }
+
+  async apply(fromBlock: number, toBlock: number): Promise<boolean> {
+    if (!this.chainId) {
+      this.logger.info('[UniswapPoolScanner] chainId not set');
+      return false;
+    }
+
+    this.uniswapV3Factory.setProvider(this.chainId);
+
+    if (!this.uniswapV3Factory.contractId) {
+      this.logger.info('[UniswapPoolScanner] uniswapV3Factory not set');
+      return false;
+    }
+
+    const events = await this.getPoolCreatedEvents(fromBlock, toBlock);
+    if (events.length == 0) return true;
+
+    const addresses = events.map((e) => [e.token1, e.token0]).flat();
+    const sortedEventsByToken = this.sortArrayByTokens(events);
+
+    for (let i = 0; i < sortedEventsByToken.length; i++) {
+      try {
+        const event = sortedEventsByToken[i];
+        const fee = Number(event.fee);
+        const pool = Number(event.pool);
+        this.logger.info(`[UniswapPoolScanner] Saving token0: ${event.token0}, token1: ${event.token1}. From ${pool}`);
+        await this.dexPoolScannerService.upsert({
+          chainId: this.chainId,
+          token0: event.token0.toLocaleLowerCase(),
+          token1: event.token1.toLocaleLowerCase(),
+          pool,
+          fee,
+        });
+      } catch (e) {
+        this.logger.error(`[UniswapPoolScanner] Could not get a valid symbol for addresses ${addresses}`);
+      }
+    }
+
+    return true;
+  }
+
+  private async getPoolCreatedEvents(fromBlock: number, toBlock: number): Promise<UniswapPoolCreatedEvent[]> {
+    this.logger.info(`[UniswapPoolScanner] Scanning for Uniswap events between ${fromBlock} to ${toBlock}`);
+    const events = await this.uniswapV3Factory.getPoolCreatedEvents(fromBlock, toBlock);
+    this.logger.info(`[UniswapPoolScanner] Got ${events.length} new Uniswap pools.`);
+    this.logger.debug(`[UniswapPoolScanner] Events: ${JSON.stringify(events)}`);
+    return events;
+  }
+
+  private sortArrayByTokens(events: PoolCreatedEvent[]) {
+    return events.sort((a, b) => {
+      const tokenA0 = a.token0.toLowerCase();
+      const tokenB0 = b.token0.toLowerCase();
+
+      if (tokenA0 === tokenB0) {
```

Comment:
this is never the case, lat's throw Error if this happen

Reply:
Done, but I'm not sure if the sort is correct tbh, if you give one example with more details?

Reply:
I just checked again, and it happened. But the logic is wrong because they could be the same token0 in the next sort iteration. I will update to use that validation:
```
      const tokenA0 = a.token0.toLowerCase();
      const tokenA1 = a.token1.toLowerCase();
      const tokenB0 = b.token0.toLowerCase();
      const tokenB1 = b.token1.toLowerCase();

      if (tokenA0 === tokenA1 || tokenB0 === tokenB1) {
        throw new Error(`${this.logPrefix} tokens are the same, pool: ${a.pool}`);
      }
```